### PR TITLE
geth: implement sync service spec

### PR DESCRIPTION
**Pull request by tynes**
**Description**
Implements the `SyncService` spec, still WIP
https://github.com/ethereum-optimism/specs/blob/main/l2-geth/transaction-ingestor.md

The spec was designed to prevent double ingestion or skipping of ingestion of transactions. The function names in the spec should match 1:1 the methods on the `SyncService`. The actual logic implemented differs slightly from the spec in how it implements its break conditions, but is functionally equivalent.

- The sequencer queries for transaction batches and compares the individual transactions to what it has locally (first step to reorgs based on L1)
- The verifier syncs using transaction batches
- Query param based calls to the DTL to query for L1 or L2 transactions (explicit instead of implicit DTL config)
- ~~Application level attempt to allow for 0 downtime rollovers of the sequencer. I think this should code should be deleted and moved to the infra layer~~

**Additional context**
This is the first step to reorgs based on L1 as well as 0 downtime rollovers
This PR builds on top of https://github.com/ethereum-optimism/optimism/pull/479 but is backwards compatible enough for the integration tests to work without it. The verifier really only takes advantage of the query param based syncing currently

### CODE REVIEW

File: l2geth/rollup/sync_service.go [hunk 1]
```diff
@@ -680,13 +577,6 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
 		return fmt.Errorf("invalid transaction: %w", err)
 	}
 
-	if tx.L1Timestamp() == 0 {
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:683]: This needs to be added back in

File: l2geth/internal/ethapi/api.go [hunk 2]
```diff
@@ -1914,8 +1914,9 @@ type EthContext struct {
 	Timestamp   uint64 `json:"timestamp"`
 }
 type RollupContext struct {
-	Index      uint64 `json:"index"`
-	QueueIndex uint64 `json:"queueIndex"`
+	Index         uint64 `json:"index"`
+	QueueIndex    uint64 `json:"queueIndex"`
+	VerifiedIndex uint64 `json:"verifiedIndex"`
```
Comment (tynes) [inline l2geth/internal/ethapi/api.go:1919]: The `VerifiedIndex` represents the last CTC index that was processed via a batch

File: l2geth/rollup/client.go [hunk 3]
```diff
@@ -90,11 +84,13 @@ type decoded struct {
 type RollupClient interface {
 	GetEnqueue(index uint64) (*types.Transaction, error)
 	GetLatestEnqueue() (*types.Transaction, error)
-	GetTransaction(index uint64) (*types.Transaction, error)
-	GetLatestTransaction() (*types.Transaction, error)
-	GetEthContext(index uint64) (*EthContext, error)
+	GetTransaction(uint64, string) (*types.Transaction, error)
+	GetLatestTransaction(string) (*types.Transaction, error)
```
Comment (tynes) [inline l2geth/rollup/client.go:88]: The `string` argument here is used as a query param sent to the DTL to specify the data source - `l1` or `l2`, see https://github.com/ethereum-optimism/optimism/pull/479

Comment (gakonst) [reply l2geth/rollup/client.go:88]: This should probably be an enum type instead of a string?

File: l2geth/internal/ethapi/api.go (hunk 2 above)
Comment (gakonst) [reply l2geth/internal/ethapi/api.go:1919]: You could add the comment in the code!

File: l2geth/rollup/client.go (hunk 3 above)
Comment (tynes) [reply l2geth/rollup/client.go]: This was updated to use an enum type

File: l2geth/rollup/sync_service.go (hunk 1 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:683]: Note: this does not need to be added back in, this logic is handled in `SyncService.applyTransactionToTip`

File: l2geth/internal/ethapi/api.go (hunk 2 above)
Comment (tynes) [reply l2geth/internal/ethapi/api.go:1924]: Added the comment to the code

**Comment by gakonst** [comment]
@tynes please add a changeset with a minor version bump by running "yarn changeset" and going through the interactive dialogue!

**Comment by tynes** [comment]
> please add a changeset with a minor version bump by running "yarn changeset" and going through the interactive dialogue!

I've added the changeset with a minor bump. This should be ready for review now

File: l2geth/rollup/client.go [hunk 4]
```diff
@@ -298,37 +299,40 @@ func transactionResponseToTransaction(res *TransactionResponse, signer *types.OV
...
 
-func (c *Client) GetTransaction(index uint64) (*types.Transaction, error) {
+func (c *Client) GetTransaction(index uint64, syncType SyncType) (*types.Transaction, error) {
 	str := strconv.FormatUint(index, 10)
 	response, err := c.client.R().
 		SetPathParams(map[string]string{
 			"index": str,
 		}).
+		SetQueryParams(map[string]string{
+			"source": syncType.String(),
```
Comment (gakonst) [inline l2geth/rollup/client.go:334]: Nit: The spec calls this `Backend`, the DTL (apparently?) calls it `source`, and geth calls it `syncType`. May want to use consistent naming across the board. No preference for which one, just that it's consistent.

File: l2geth/rollup/sync_service_test.go [hunk 5]
```diff
@@ -48,7 +48,7 @@ func TestSyncServiceContextUpdated(t *testing.T) {
 	}
 
 	// run the update context call once
-	err := service.updateContext()
+	err := service.updateEthContext()
```
Comment (gakonst) [inline l2geth/rollup/sync_service_test.go:51]: This rename should also ideally be in a separate PR.

File: l2geth/rollup/sync_service.go [hunk 6]
```diff
@@ -75,8 +72,8 @@ func NewSyncService(ctx context.Context, cfg Config, txpool *core.TxPool, bc *co
 	}
 	timestampRefreshThreshold := cfg.TimestampRefreshThreshold
 	if timestampRefreshThreshold == 0 {
-		log.Info("Sanitizing timestamp refresh threshold to 15 minutes")
-		timestampRefreshThreshold = time.Minute * 15
+		log.Info("Sanitizing timestamp refresh threshold to 5 minutes")
+		timestampRefreshThreshold = time.Minute * 5
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:76]: Any reason why this default was changed?

File: l2geth/rollup/sync_service.go [hunk 7]
```diff
@@ -86,24 +83,33 @@ func NewSyncService(ctx context.Context, cfg Config, txpool *core.TxPool, bc *co
 	}
 	// Initialize the rollup client
 	client := NewClient(cfg.RollupClientHttp, chainID)
-	log.Info("Configured rollup client", "url", cfg.RollupClientHttp, "chain-id", chainID.Uint64(), "ctc-deploy-height", cfg.CanonicalTransactionChainDeployHeight)
+	log.Info("Configured rollup client", "url", cfg.RollupClientHttp, "chain-id", chainID.Uint64())
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:86]: Should we not keep this log as was before?

File: l2geth/rollup/sync_service.go [hunk 8]
```diff
@@ -86,24 +83,33 @@ func NewSyncService(ctx context.Context, cfg Config, txpool *core.TxPool, bc *co
...
 	service := SyncService{
 		ctx:                       ctx,
 		cancel:                    cancel,
 		verifier:                  cfg.IsVerifier,
 		enable:                    cfg.Eth1SyncServiceEnable,
-		confirmationDepth:         cfg.Eth1ConfirmationDepth,
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:95]: Why is it OK to remove this?

File: l2geth/rollup/sync_service.go [hunk 9]
```diff
@@ -86,24 +83,33 @@ func NewSyncService(ctx context.Context, cfg Config, txpool *core.TxPool, bc *co
...
 	}
 
-	// Initial sync service setup if it is enabled. This code depends on
+	// The chainHeadSub is used to synchronize the SyncService with the chain.
+	// As the SyncService processes transactions, it waits until the transaction
+	// is added to the chain. This synchronization is required for handling
+	// reorgs and also favors safety over liveliness. If a transaction breaks
+	// things downstream, it is expected that this channel will halt ingestion
+	// of additional transactions by the SyncService.
+	service.chainHeadSub = service.bc.SubscribeChainHeadEvent(service.chainHeadCh)
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:110]: I don't get it - how does this sub help? I don't see it being used anywhere?

File: l2geth/rollup/sync_service.go [hunk 10]
```diff
@@ -303,371 +305,308 @@ func (s *SyncService) VerifierLoop() {
...
-		log.Debug("Applying transaction", "index", i)
-		err = s.maybeApplyTransaction(tx)
+	case SyncTypeSequenced:
+		err := s.syncTransactionsToTip(SyncTypeSequenced)
 		if err != nil {
-			return fmt.Errorf("could not apply transaction: %w", err)
+			log.Error("Verifier cannot sync transactions with SyncTypeSequencer", "msg", err)
 		}
-		s.SetLatestIndex(&i)
 	}
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:322]: This switch logic seems error prone to me? 

Seems better for us to have a `ReplicaSyncService` and a `L1SyncService` and have them internally implement the required logic for each of the functionalities explicitly.

File: l2geth/rollup/sync_service_test.go [hunk 11]
```diff
@@ -217,11 +212,7 @@ func TestSyncServiceSync(t *testing.T) {
 		},
 	})
 
-	err = service.verify()
-	if err != nil {
-		t.Fatal("verification failed", err)
-	}
-
+	go service.syncTransactionsToTip(SyncTypeSequenced)
```
Comment (gakonst) [inline l2geth/rollup/sync_service_test.go:215]: Same as above comment on goroutine

File: l2geth/rollup/sync_service_test.go [hunk 12]
```diff
@@ -132,11 +132,7 @@ func TestSyncServiceTransactionEnqueued(t *testing.T) {
 		},
 	})
 
-	// Run an iteration of the eloop
-	err = service.sequence()
-	if err != nil {
-		t.Fatal("sequencing failed", err)
-	}
+	go service.syncQueueToTip()
```
Comment (gakonst) [inline l2geth/rollup/sync_service_test.go:135]: Why was this changed? IMO it is super error prone to use go-routines which we don't know how many iterations they'll run for inside tests. Better to explicitly run a single iteration (or N iterations for N transactions) inside.

File: l2geth/rollup/sync_service.go [hunk 13]
```diff
@@ -755,6 +853,7 @@ func getRawTransaction(tx *types.Transaction) ([]byte, error) {
 	return data.Bytes(), nil
 }
 
+// fillBytes is taken from a newer version of the golang standard library
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:856]:
```suggestion
// fillBytes is taken from a newer version of the golang standard library
// will panic if the provided biginteger does not fit in the provided `size` number of bytes
```

File: l2geth/rollup/sync_service.go [hunk 14]
```diff
@@ -704,10 +636,176 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
...
+		if postLatestIndex == nil {
+			return errors.New("Latest index is nil")
+		}
+		if *postLatestIndex == *latestIndex {
+			return nil
+		}
+	}
+}
+
+// syncTransactionBatchesToTip will sync all of the transaction batches to the
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:694]: Recommendation: Write the algorithm being implemented in the doc comment. 

Example:
1. It gets the latest transaction batch index from the DTL
2. It gets the current batch index from the database
3. It gets any batches between the two, and applies all the transactions inside that batch sequentially

File: l2geth/rollup/sync_service.go [hunk 15]
```diff
@@ -704,10 +636,176 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
...
+		nextIndex := s.GetNextVerifiedIndex()
+
+		for i := nextIndex; i <= latestIndex; i++ {
+			log.Debug("Fetching transaction batch", "index", i)
+			_, txs, err := s.client.GetTransactionBatch(i)
+			if err != nil {
+				return fmt.Errorf("Cannot get transaction batch: %w", err)
+			}
+			for _, tx := range txs {
+				s.applyBatchedTransaction(tx)
```
Comment (gakonst) [inline l2geth/rollup/sync_service.go:720]: You are not handling the error here.

**Review by gakonst — CHANGES_REQUESTED** [review]
A few changes requested internally. 

Also, as a process point, I think we should be more diligent in having explicit commits for each change being made inside geth and contracts, since they are critical components for the core protocol.

This PR does a lot of refactoring/renaming, when that should be saved for a separate PR which changes no functionality and simply moves code around.

As another example, [this commit](https://github.com/ethereum-optimism/optimism/pull/477/commits/8599671ee091917edff8745c8db5960355e5a109) should have been split into a few commits:
* `refactor: use transaction instead of response when parsing dtl response`
*  `refactor: rename updateContext to updateEthContext`
* `feat: allow reading and writing verified batch indexes to the db`
* `feat: add backend parameter to DTL queries`

Similarly, when removing stuff, e.g. `confirmationDepth`, you can add a commit that says `chore: remove unused confirmationDepth param`.

Finally, if the DTL "choose your own backend" changes are NOT needed to implement the basic functionality, then they should be split in a separate PR. It does not make sense to add things to `client.go` in this PR if they are not present in the DTL as well.

OK if you don't want to do this for this PR, but it'll make reviewing such code changes much easier in the future.

Another tip would be that after you're doing with your PR, `git reset master`, and then start re-doing the commits individually so that they're logically split with `git add -p`, to stage the hunks that matter for each logical change.

"fix typos" and "fix broken tests" commits are not useful for the reviewer, since they are changes introduced in this PR. The commit message is not about things you've done related to your previous commits, it's related to the things you've done related to master!

File: l2geth/rollup/client.go (hunk 4 above)
Comment (tynes) [reply l2geth/rollup/client.go:334]: The dichotomy is "batch submitted" or "not batch submitted". Its not necessarily L1 or L2 as backends because what is L1 today could be ETH2 shard data in the future. We do need a unified language for this

File: l2geth/rollup/sync_service.go (hunk 6 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:76]: 15 minutes with no timestamp updates is way too long for users. It makes batch submission slightly cheaper due to less contexts but not worth the degraded UX

File: l2geth/rollup/sync_service.go (hunk 7 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:86]: The `ctc-deploy-height` isn't really relevant to the `RollupClient`. I figured it makes sense to group things contextually, it is not logged inside of `initializeLatestL1`

File: l2geth/rollup/sync_service.go (hunk 10 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:322]: The `SyncService.syncType` is not dynamic and is configured up front. You are worried about misconfiguration? 

We would need `SequencerSyncService`, `ReplicaSyncService` and `L1SyncService` - they share a lot of functionality and breaking them out would add a few hundred lines of code. I'd definitely prefer to not break them apart

File: l2geth/rollup/sync_service.go [hunk 16]
```diff
@@ -303,371 +305,308 @@ func (s *SyncService) VerifierLoop() {
...
+	// This is a temporary fix for a bug in the SequencerEntrypoint. It will
+	// be removed when the custom batch serialization is removed in favor of
+	// batching RLP encoded transactions.
 	tx = fixType(tx)
+
 	txs := types.Transactions{tx}
 	s.txFeed.Send(core.NewTxsEvent{Txs: txs})
+	// Block until the transaction has been added to the chain
+	log.Debug("Waiting for transaction to be added to chain", "hash", tx.Hash().Hex())
+	<-s.chainHeadCh
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:569]: @gakonst The `chainHeadCh` is used here. This blocks until the the `tx` is added to the chain. Without this, events are sent over the `txFeed` much faster than they can be processed and things start breaking due to the combination of `Clique` and the miner being used

File: l2geth/rollup/sync_service.go (hunk 15 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:720]: Good catch

File: l2geth/rollup/sync_service.go (hunk 14 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:694]: I'd like to update the spec with this algorithm - the name of the function is `syncTransactionBatchesToTip` and with the outer `while true` it is able to catch new batches that are submitted while it has been processing. The logic in the spec has assumptions that syncing to the tip actually does syncing to the tip, so in practice this handles the race condition

File: l2geth/rollup/sync_service_test.go (hunk 12 above)
Comment (tynes) [reply l2geth/rollup/sync_service_test.go:135]: `syncQueueToTip` does less things than `sequence` and calling it once is a single iteration. This starts a goroutine that runs until `syncQueueToTip` returns. This was added because if it does not run in a goroutine, it blocks due to the addition of `txCh` and waiting explicitly for transactions to be added to the blockchain after processing them

File: l2geth/internal/ethapi/api.go [hunk 17]
```diff
@@ -1913,9 +1913,15 @@ type EthContext struct {
...
+// RollupContext represents the height of the rollup.
+// Index is the last processed CanonicalTransactionChain index
+// QueueIndex is the last processed `enqueue` index
+// VerifiedIndex is the last processed CTC index that was batched
 type RollupContext struct {
-	Index      uint64 `json:"index"`
-	QueueIndex uint64 `json:"queueIndex"`
+	Index         uint64 `json:"index"`
+	QueueIndex    uint64 `json:"queueIndex"`
+	VerifiedIndex uint64 `json:"verifiedIndex"`
```
[the remark this answers is not in the document]
Comment (tynes) [reply l2geth/internal/ethapi/api.go:1924]: `BatchIndex` is different from `VerifiedIndex` - the `VerifiedIndex` is still a CTC index while a `BatchIndex` is the n'th batch that has been submitted

File: l2geth/eth/api_backend.go [hunk 18]
```diff
@@ -318,7 +323,7 @@ func (b *EthAPIBackend) SendTx(ctx context.Context, signedTx *types.Transaction)
 				return fmt.Errorf("Calldata cannot be larger than %d, sent %d", b.MaxCallDataSize, len(signedTx.Data()))
 			}
 		}
-		return b.eth.syncService.ApplyTransaction(signedTx)
+		return b.eth.syncService.ValidateAndApplySequencerTransaction(signedTx)
```
[the remark this answers is not in the document]
Comment (tynes) [reply l2geth/eth/api_backend.go:326]: I was copying the spec and there was strong opinions around naming the function that in the spec. I'm not super opinionated on this particular topic, but some functions assume validation has already happened

File: l2geth/rollup/sync_service.go (hunk 14 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:694]: https://github.com/ethereum-optimism/specs/pull/31

File: l2geth/rollup/client.go [hunk 19]
```diff
@@ -222,45 +223,45 @@ func (c *Client) GetLatestEnqueue() (*types.Transaction, error) {
...
 	// `nil` transactions are not found
-	if res.Transaction == nil {
+	if res == nil {
 		return nil, nil
 	}
 	// The queue origin must be either sequencer of l1, otherwise
 	// it is considered an unknown queue origin and will not be processed
 	var queueOrigin types.QueueOrigin
-	if res.Transaction.QueueOrigin == "sequencer" {
+	if res.QueueOrigin == "sequencer" {
```
[the remark this answers is not in the document]
Comment (tynes) [reply l2geth/rollup/client.go:234]: https://github.com/ethereum-optimism/optimism/pull/516

File: l2geth/eth/api_backend.go [hunk 20]
```diff
@@ -71,9 +71,10 @@ func (b *EthAPIBackend) GetEthContext() (uint64, uint64) {
 	return bn, ts
 }
 
-func (b *EthAPIBackend) GetRollupContext() (uint64, uint64) {
+func (b *EthAPIBackend) GetRollupContext() (uint64, uint64, uint64) {
 	i := uint64(0)
 	q := uint64(0)
+	v := uint64(0)
```
[the remark this answers is not in the document]
Comment (tynes) [reply l2geth/eth/api_backend.go:77]: https://github.com/ethereum-optimism/optimism/pull/520

File: l2geth/rollup/sync_service.go (hunk 6 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:76]: Setting it to 3 minutes https://github.com/ethereum-optimism/optimism/pull/531

**Comment by tynes** [comment]
Closed in favor of https://github.com/ethereum-optimism/optimism/pull/552
